base/strings/string.jl

```diff
@@ -81,8 +81,8 @@ String(s::CodeUnits{UInt8,String}) = s.s
 pointer(s::String) = unsafe_convert(Ptr{UInt8}, s)
 pointer(s::String, i::Integer) = pointer(s)+(i-1)
 
-ncodeunits(s::String) = Core.sizeof(s)
-sizeof(s::String) = Core.sizeof(s)
+@pure ncodeunits(s::String) = Core.sizeof(s)
+@pure sizeof(s::String) = Core.sizeof(s)
```

Comment:
I think inference is supposed to know this one already, and might just need a minor fix

Reply:
Actually a while ago I specifically told the compiler not to constant prop Strings since it led to extremely long compile times for some packages (if I recall, DiffEq parsing many string literals to input numeric constants).

Reply:
purity detection should be able to infer this independently of constant prop though and mark it as pure